file-upload/file-upload-scala-sbt/fileupload-impl/src/main/scala/play/api/routing/AkkaHttpPlayRouter.scala

```diff
@@ -0,0 +1,33 @@
+package play.api.routing
+
+import akka.actor.ActorSystem
+import akka.http.scaladsl.model.{HttpRequest, HttpResponse}
+import akka.http.scaladsl.server.{Route, RoutingLog}
+import akka.http.scaladsl.settings.{ParserSettings, RoutingSettings}
+import akka.stream.{ActorMaterializerHelper, Materializer}
+import play.api.mvc.akkahttp.AkkaHttpHandler
+import play.api.routing.Router.Routes
+
+import scala.concurrent.Future
+
+object AkkaHttpPlayRouter {
+
+  def apply(prefix: String, akkaHttpRoute: Route)(implicit actorSystem: ActorSystem, materializer: Materializer): Router = {
+    val routingSettings = RoutingSettings(actorSystem)
+    val routingLog = RoutingLog.fromActorSystem(actorSystem)
+    val parserSettings = ParserSettings(actorSystem)
+    apply(prefix, Route.asyncHandler(akkaHttpRoute)(routingSettings, parserSettings, materializer, routingLog))
```

Comment:
Using `Route.asyncHandler` has the inconvenience that it internally `seals` the `Route`. As a consequence, the code needs to be:

```scala
  lazy val clusterHttpRouter = AkkaHttpPlayRouter("/cluster", ClusterHttpManagementRoutes(Cluster(ac)))
  //...
  override lazy val lagomServer: LagomServer =
    serverFor[FileUploadService](wire[FileUploadServiceImpl])
      .plusRouter(clusterHttpRouter)
      .plusRouter(fileUploadRouter.withPrefix("/io"))
```

which is quite error prone considering that:

 * `.plusRouter(fileUploadRouter.withPrefix("/io"))` does manipulate the routes to append `/io`
 * `AkkaHttpPlayRouter("/abc", xyz)` doesn't manipulate the routes in `xyz` appending `/abc` but, instead, must match it's internal top-level `pathPrefix` (in this case `/cluster`).

I remember @marcospereira and @renatocaval (you) discussing alternatives to remove prefixes before at play level before passing the request to the `Router`, I guess you already faced this...